pkg/app/ops/deploymentchaincontroller/updater.go

```diff
@@ -0,0 +1,189 @@
+			if model.IsCompletedDeployment(deploymentRef.Status) {
+				continue
+			}
+			deployment, err := u.deploymentStore.GetDeployment(ctx, deploymentRef.DeploymentId)
+			if err != nil {
+				return err
+			}
+			// Update the deployment state in deployment chain model in case
+			// the deployment state is changed.
+			if deployment.Status != deploymentRef.Status {
+				if err = u.deploymentChainStore.UpdateDeploymentChain(ctx, u.deploymentChainID,
+					datastore.DeploymentChainNodeDeploymentStatusUpdater(
+						deployment.DeploymentChainBlockIndex,
+						deployment.Id,
+						deployment.Status,
+						deployment.StatusReason),
+				); err != nil {
+					return err
+				}
+				// Update updater's deploymentRefs.
+				deploymentRef.Status = deployment.Status
+				deploymentRef.StatusReason = deployment.StatusReason
+			}
+		}
+	}
+
+	// If not all applications in chain has its deployment ref
+	// fetch all missing deployments and link the deployment ref to them.
+	if len(u.deploymentRefs) != len(u.applicationRefs) {
+		deployments, err := u.listAllMissingDeployments(ctx)
+		if err != nil {
+			return err
+		}
+
+		// Note: Only the deployment of the first block (the deployment which trigger this chain)
+		// is added at the time we created this deployment chain; thus all other deployments
+		// which be counted as "missng deployments" here are triggered and going to be added under
+		// PENDING state, since those deployments are of blocks which have to wait for its previous
+		// blocks completion. We don't need to care about updating those deployment ref status after
+		// they been added here in this interval.
+		for _, deployment := range deployments {
+			// Connect missing deployment to the chain.
+			if err := u.deploymentChainStore.UpdateDeploymentChain(
+				ctx,
+				u.deploymentChainID,
+				datastore.DeploymentChainAddDeploymentToBlock(deployment),
+			); err != nil {
+				return err
+			}
+			// Store deployment ref to the updater's deploymentRefs.
+			u.deploymentRefs[deployment.ApplicationId] = &model.ChainDeploymentRef{
+				DeploymentId: deployment.Id,
+				Status:       deployment.Status,
+				StatusReason: deployment.StatusReason,
+			}
+		}
+	}
+
+	return nil
+}
+
+func (u *updater) listAllMissingDeployments(ctx context.Context) ([]*model.Deployment, error) {
+	noDeploymentApps := make([]string, 0, len(u.applicationRefs))
+	for _, appRef := range u.applicationRefs {
+		if _, ok := u.deploymentRefs[appRef.ApplicationId]; !ok {
+			noDeploymentApps = append(noDeploymentApps, appRef.ApplicationId)
+		}
+	}
+
+	deployments := make([]*model.Deployment, 0, len(noDeploymentApps))
+	// TODO: Find a better way to fetch applications in batch.
+	for _, appID := range noDeploymentApps {
+		app, err := u.applicationStore.GetApplication(ctx, appID)
+		if err != nil {
+			return nil, err
+		}
+		// If the most recent triggered deployment does not exist, ignore it.
+		if app.MostRecentlyTriggeredDeployment == nil {
```

Comment:
Relying on the `MostRecentlyTriggeredDeployment` is not safe. Another deployment could be triggered between the checkpoints and that causes a wrong result.

How do you think about listing all deployments of the chain by filtering with the chain ID?

Reply:
You're right, I took the detour. Nice catch 👍

Reply:
Looks better 👍 Addressed by https://github.com/pipe-cd/pipe/pull/2986/commits/7b6f48bf4eb690266845f686a7fccb98cc36c6b2